### ISSUE TITLE
[R] Potential memory leak when lgb is called repetively on linux maschines

### ISSUE THREAD

**Issue by weltherrschaf** [Microsoft/LightGBM#1066]
I am doing hyperparameter optimization with lgb and thus call lgb zillions of times and noticed a small memory leak when lgb is repeatedly called. It seems to come from using the restore=T parameter in lgb.unloader. Memory can't be freed by calls to gc(), only by quitting the session. This code reproduces the error on CentOS and Ubuntu

a <- 1
counter <- 0
while(a==1){
  counter <- counter+1
  print(paste("Repetition", counter, "started at", Sys.time()))
  #library(lightgbm)
  data(agaricus.train, package = "lightgbm")
  data(agaricus.test, package = "lightgbm")
  train <- agaricus.train
  test <- agaricus.test
  dtrain <- lightgbm::lgb.Dataset(data = train$data,
                        label = train$label)
  bst <- lightgbm::lightgbm(data = dtrain,
                  num_leaves = 4,
                  learning_rate = 1,
                  nrounds = 2,
                  objective = "binary", verbose = -1, num_threads = 1)
  lightgbm::lgb.unloader(wipe = TRUE, restore = TRUE)
}

when using restore=F, memory isn't accumulated. Thus as an intermediate solution I can call lgb via namespace.

sessionInfo()

R version 3.4.1 (2017-06-30)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] lightgbm_2.0.10 Matrix_1.2-10   R6_2.2.2

loaded via a namespace (and not attached):
[1] compiler_3.4.1      magrittr_1.5        tools_3.4.1
[4] grid_3.4.1          data.table_1.10.4-3 lattice_0.20-35

I tried it too with a ubuntu 16.x system sporting the latest R 3.4.2..

**Comment by Laurae2** [comment]
`lgb.unloader` is a convenience function for when the user does not know why LightGBM does not work after an error, I recommend doing this way as it does not increase memory usage:

```r
a <- 1
counter <- 0
data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")
library(lightgbm)
while(a==1){
  counter <- counter+1
  print(paste("Repetition", counter, "started at", Sys.time()))
  train <- agaricus.train
  test <- agaricus.test
  dtrain <- lgb.Dataset(data = train$data,
                                  label = train$label)
  bst <- lightgbm(data = dtrain,
                            num_leaves = 4,
                            learning_rate = 1,
                            nrounds = 2,
                            objective = "binary", verbose = -1, num_threads = 1)
  rm(train, test, dtrain, bst)
  gc(verbose = FALSE)
}
```

Also check this example: https://github.com/Microsoft/LightGBM/blob/master/R-package/demo/efficient_many_training.R
